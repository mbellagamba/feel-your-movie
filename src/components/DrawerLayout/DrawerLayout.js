import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { useLocation } from 'react-router-dom';
import { TEXT_PRIMARY } from '../../resources/colors';
import { MARGIN_MEDIUM, TEXT_MEDIUM } from '../../resources/dimensions';
import MenuIcon from '../../../images/menu.svg';

import MenuItem from '../MenuItem';

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100%;
  color: ${TEXT_PRIMARY};
`;

const Toolbar = styled.header`
  height: 4rem;
  border-bottom: 1px solid white;
  display: flex;
  flex: 0 0 auto;
  flex-flow: row nowrap;
  align-items: center;
  padding-left: ${MARGIN_MEDIUM};
  padding-right: ${MARGIN_MEDIUM};
`;

const Body = styled.div`
  display: flex;
  flex-direction: row;
  flex: 1 1 auto;
  position: relative;
  overflow-y: auto;
  margin-top: ${MARGIN_MEDIUM};
`;

const Sidebar = styled.div`
  display: flex;
  font-size: ${TEXT_MEDIUM};
  width: ${(props) => (props.visible ? '360px' : '0')};
  flex-direction: column;
  transition: width .25s ease-in-out;
`;

const Main = styled.div`
  width: 100%;
  margin-left: ${MARGIN_MEDIUM};
  margin-right: ${MARGIN_MEDIUM};
`;

const Title = styled.h2`
  margin: 0px ${MARGIN_MEDIUM};
`;

const DrawerLayout = ({ children }) => {
  const { pathname } = useLocation();
  const [menuOpen, setMenuOpen] = useState(true);

  const handleClick = (event) => {
    event.preventDefault();
    setMenuOpen(!menuOpen);
  };

  return (
    <Container>
      <Toolbar>
        <button type="button" onClick={handleClick}>
          <img src={MenuIcon} alt="Menu" />
        </button>
        <Title>Feel your movie</Title>
      </Toolbar>
      <Body>
        <Sidebar visible={menuOpen}>
          <MenuItem title="Home" path="/" currentPath={pathname} />
          <MenuItem title="Movies" path="/movies" currentPath={pathname} />
          <MenuItem title="Suggestion" path="/suggestion" currentPath={pathname} />
          <MenuItem title="About" path="/about" currentPath={pathname} />
        </Sidebar>
        <Main>
          {children}
        </Main>
      </Body>
    </Container>
  );
};

DrawerLayout.propTypes = {
  children: PropTypes.node.isRequired,
};

export default DrawerLayout;
