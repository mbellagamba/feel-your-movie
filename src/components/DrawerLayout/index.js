import React, { useState } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { useLocation, Link } from 'react-router-dom';
import { useWindowDimensions } from '../../hooks';
import {
  TEXT_PRIMARY,
  ACCENT,
  BACKGROUND,
} from '../../resources/colors';
import {
  MARGIN_MEDIUM,
  TEXT_MEDIUM,
  TOOLBAR_HEIGHT,
  SCREEN_WIDTH_PHONE,
  TEXT_SMALL,
} from '../../resources/dimensions';
import menu from '../../../images/menu.svg';
import home from '../../../images/home.svg';
import movie from '../../../images/movie.svg';
import lightbulb from '../../../images/lightbulb.svg';
import info from '../../../images/info.svg';

const menuItems = [
  {
    title: 'Home',
    path: '/',
    icon: home,
  },
  {
    title: 'Movies',
    path: '/movies',
    icon: movie,
  },
  {
    title: 'Suggestion',
    path: '/suggestion',
    icon: lightbulb,
  },
  {
    title: 'About',
    path: '/about',
    icon: info,
  },
];

const Container = styled.div`
  color: ${TEXT_PRIMARY};
`;

const Toolbar = styled.header`
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  z-index: 1;
  height: ${TOOLBAR_HEIGHT};
  border-bottom: 1px solid white;
  display: flex;
  flex-flow: row nowrap;
  align-items: center;
  background-color: ${BACKGROUND};
  padding-left: ${MARGIN_MEDIUM};
  padding-right: ${MARGIN_MEDIUM};
`;

const Body = styled.div`
  display: flex;
  flex-direction: row;
  flex: 1 1 auto;
  position: relative;
  overflow-y: auto;
  margin-top: ${TOOLBAR_HEIGHT};
`;

const Sidebar = styled.div`
  display: flex;
  font-size: ${TEXT_MEDIUM};
  margin-top: ${MARGIN_MEDIUM};
  width: ${(props) => (props.visible ? '260px' : '0')};
  overlow-x: hidden;
  flex-direction: column;
  transition: width .25s ease-out;
  @media (max-width: ${SCREEN_WIDTH_PHONE}) {
    width: ${(props) => (props.visible ? '80%' : '0')}
  }
`;

const Main = styled.div`
  flex: 1;
  margin-left: ${MARGIN_MEDIUM};
  margin-right: ${MARGIN_MEDIUM};
  @media (max-width: ${SCREEN_WIDTH_PHONE}) {
    display: ${(props) => (props.menuOpen ? 'none' : 'block')}
  }
`;

const Title = styled.h2`
  margin: 0px ${MARGIN_MEDIUM};
`;

const MenuButton = styled.button`
  padding: 0;
`;

const MenuItem = styled.div`
  display: flex;
  flex-direction: row;
  line-height: 3rem;
  font-size: ${TEXT_SMALL}
  border-top-right-radius: 1.5rem;
  border-bottom-right-radius: 1.5rem;
  overflow: hidden;
  ${(props) => (props.active
    ? `background-color: ${ACCENT}40;`
    : ':hover { background-color: #ffffff22; }')}
`;

const Icon = styled.img`
  margin: 0 ${MARGIN_MEDIUM};
`;

const DrawerLayout = ({ children }) => {
  const { pathname } = useLocation();
  const { width } = useWindowDimensions();
  const desktop = width > parseInt(SCREEN_WIDTH_PHONE, 10);
  const [menuOpen, setMenuOpen] = useState(desktop);

  const handleClick = (event) => {
    event.preventDefault();
    setMenuOpen(!menuOpen);
  };

  const onMenuItemClick = () => {
    if (!desktop) {
      setMenuOpen(false);
    }
  };

  const isActive = (path, current) => {
    if (!current) return false;
    return pathname === path || (path !== '/' && pathname.startsWith(path));
  };

  return (
    <Container>
      <Toolbar>
        <MenuButton type="button" onClick={handleClick}>
          <img src={menu} alt="Menu" />
        </MenuButton>
        <Title>Feel your movie</Title>
      </Toolbar>
      <Body>
        <Sidebar data-testid="sidebar" visible={menuOpen}>
          {menuItems.map(({ title, path, icon }) => (
            <Link key={title} to={path} onClick={onMenuItemClick}>
              <MenuItem active={isActive(path, pathname)}>
                <Icon src={icon} alt={title} />
                {title}
              </MenuItem>
            </Link>
          ))}
        </Sidebar>
        <Main menuOpen={menuOpen}>
          {children}
        </Main>
      </Body>
    </Container>
  );
};

DrawerLayout.propTypes = {
  children: PropTypes.node.isRequired,
};

export default DrawerLayout;
