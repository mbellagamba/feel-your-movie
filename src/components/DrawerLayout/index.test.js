import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import DrawerLayout from './index';

jest.mock('react-router-dom', () => ({
  Link: () => <div />,
  useLocation: jest.fn(() => 'fake-route'),
}));

jest.mock('../../hooks', () => ({
  useWindowDimensions: jest.fn(() => ({ width: 1024 })),
}));

test('should close the sidebar on menu icon click', () => {
  const { getByTestId } = render(<DrawerLayout><div /></DrawerLayout>);
  const sidebar = getByTestId('sidebar');
  expect(sidebar).toHaveStyleRule('width', '260px');
  fireEvent.click(getByTestId('menu-button'));
  expect(sidebar).toHaveStyleRule('width', '0');
});

test('should match snapshot', () => {
  const { asFragment } = render(<DrawerLayout><div /></DrawerLayout>);
  expect(asFragment()).toMatchSnapshot();
});
