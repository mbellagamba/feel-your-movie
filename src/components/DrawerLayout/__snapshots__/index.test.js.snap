// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  color: #fff;
}

.c1 {
  height: 4rem;
  border-bottom: 1px solid white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 16px;
  padding-right: 16px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  position: relative;
  overflow-y: auto;
  margin-top: 16px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 1.2rem;
  width: 360px;
  overlow-x: hidden;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-transition: width .25s ease-in-out;
  transition: width .25s ease-in-out;
}

.c6 {
  width: 100%;
  margin-left: 16px;
  margin-right: 16px;
}

.c3 {
  margin: 0px 16px;
}

.c2 {
  padding: 0;
}

@media (max-width:400px) {
  .c6 {
    display: none;
  }
}

<div
    class="c0"
  >
    <header
      class="c1"
    >
      <button
        class="c2"
        type="button"
      >
        <img
          alt="Menu"
          src="test-file-stub"
        />
      </button>
      <h2
        class="c3"
      >
        Feel your movie
      </h2>
    </header>
    <div
      class="c4"
    >
      <div
        class="c5"
        data-testid="sidebar"
      >
        <div />
        <div />
        <div />
        <div />
      </div>
      <div
        class="c6"
      >
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
